src/converters/intrinsics.ts

```diff
@@ -251,6 +313,130 @@ export const fnEquals: Intrinsic = {
+            return ctx.getRegion();
+        case 'AWS::URLSuffix':
+            return ctx.getURLSuffix();
+        case 'AWS::NotificationARNs':
+        case 'AWS::StackId':
+        case 'AWS::StackName': {
+            // TODO[pulumi/pulumi-cdk#246]: these pseudo-parameters are typically used in things like names or descriptions
+            // so it should be safe to substitute with a stack node ID for most applications.
+            const stackNodeId = ctx.getStackNodeId();
+            debug(`pulumi-cdk is replacing a Ref to a CF pseudo-parameter ${param} with the stack node ID`);
+            return stackNodeId;
+        }
+    }
+
+    // Handle Cf template parameters.
+    const cfParam = ctx.findParameter(param);
+    if (cfParam !== undefined) {
+        return ctx.evaluateParameter(cfParam);
+    }
+
+    // Handle references to resources.
+    const map = ctx.findResourceMapping(param);
+    if (map !== undefined) {
+        if (map.attributes && 'id' in map.attributes) {
+            // Users my override the `id` in a custom-supplied mapping, respect this.
+            return ctx.succeed(map.attributes.id);
+        }
+        if (aws.cloudformation.CustomResourceEmulator.isInstance(map.resource)) {
+            // Custom resources have a `physicalResourceId` that is used for Ref
+            return ctx.succeed(map.resource.physicalResourceId);
+        }
+        const resMeta = ctx.tryFindResource(map.resourceType);
+
+        // If there is no metadata to suggest otherwise, assume that we can use the Pulumi id which typically will be
+        // the primaryIdentifier from CloudControl.
+        if (resMeta === undefined || !resMeta.cfRef || resMeta.cfRef.notSupportedYet) {
+            const cr = <pulumi.CustomResource>map.resource; // assume we have a custom resource.
+            return ctx.succeed(cr.id);
+        }
+
+        // Respect metadata if it suggests Ref is not supported.
+        if (resMeta.cfRef.notSupported) {
+            return ctx.fail(`Ref intrinsic is not supported for the ${map.resourceType} resource type`);
+        }
+
+        // At this point metadata should indicate which properties to extract from the resource to compute the ref.
+        const propNames: string[] = (resMeta.cfRef.properties || [])
+            .concat(resMeta.cfRef.property ? [resMeta.cfRef.property] : [])
+            .map((x) => toSdkName(x));
+
+        const propValues: any[] = [];
+        for (const p of propNames) {
+            if (!Object.prototype.hasOwnProperty.call(map.resource, p)) {
+                return ctx.fail(`Pulumi metadata notes a property "${p}" but no such property was found on a resource`);
+            }
+            propValues.push((<any>map.resource)[p]);
+        }
+
+        const delim: string = resMeta.cfRef!.delimiter || '|';
+
+        return ctx.apply(ctx.succeed(propValues), (resolvedValues) => {
+            let i = 0;
+            for (const v of resolvedValues) {
```

Comment:
Nit: 

```ts
resolvedValues.forEach((v, idx) => {
	...
})
```

Reply:
forEach wouldn't have the early return. To express this functionally this would be a reduce, but perhaps having written too much Go since my FP days I prefer the imperative version now for readability.